src/Build/Graph/ProjectGraph.cs

```diff
@@ -75,6 +76,69 @@ public sealed class ProjectGraph
 
         public IReadOnlyCollection<ProjectGraphNode> GraphRoots { get; }
 
+        /// <summary>
+        ///     Gets a distinct enumeration of all of the environment variables reads during evaluation.
+        /// </summary>
+        public IEnumerable<string> EnvironmentVariableReads
+        {
+            get
+            {
+                if (!ProjectNodes.Any())
+                {
+                    return Enumerable.Empty<string>();
+                }
+
+                var distinctReads = new HashSet<string>();
+
+                foreach (ProjectGraphNode projectNode in ProjectNodes)
+                {
+                    foreach (string environmentVariableRead in projectNode.ProjectInstance.EnvironmentVariableReads)
+                    {
+                        if (!distinctReads.Contains(environmentVariableRead))
```

Comment:
You can skip the Contains check and just do add, saving a lookup.

Reply:
same below

Reply:
Updated.